File: sdk/typescript/kafka/kafkaConsumer.integration.test.ts
/**
 * Integration test for the real kafkajs adapter against a LIVE broker.
 *
 * GATED: only runs when KAFKA_BOOTSTRAP is set. Without it the whole suite is
 * skipped so the normal `node --test` unit run stays broker-free / green.
 *
 *   KAFKA_BOOTSTRAP=localhost:19092 \
 *     node --test --experimental-strip-types 'kafka/*.integration.test.ts'
 *
 * Coverage (byo-sdk-guide §1.2 / §1.9, wire-protocol §A):
 *   - admin creates `batch.task.dispatch.acme.<chan>` (unique suffix)
 *   - adapter (tenantId 'acme') subscribes via the wildcard regex + manual commit
 *   - a valid v1 TaskDispatchMessage is RECEIVED + ACCEPTED (committed)
 *   - a foreign-tenant message is DROPPED (not committed)
 *   - a schemaVersion "v3" message is REJECTED (not committed)
 *
 * Cleanup: unique topic/group per run; topics deleted in a finally block.
 */

import { test } from "node:test";
import assert from "node:assert/strict";

import {
  MessagePipeline,
  assignmentOf,
  type DispatchMessage,
  type PipelineOutcome,
} from "../src/consumer.ts";
import {
  KafkaConsumerAdapter,
  consumerGroupId,
  dispatchTopicRegex,
} from "./kafkaConsumer.ts";

const BOOTSTRAP = process.env.KAFKA_BOOTSTRAP;

const silentLogger = { info: () => {}, warn: () => {}, error: () => {} };

/** Wait until `predicate()` is true or time runs out. */
async function waitFor(
  predicate: () => boolean,
  timeoutMs = 20_000,
  stepMs = 100,
): Promise<void> {
  const deadline = Date.now() + timeoutMs;
  while (Date.now() < deadline) {
    if (predicate()) return;
    await new Promise((r) => setTimeout(r, stepMs));
  }
  throw new Error("waitFor: timed out");
}

test(
  "kafka integration: accept v1, drop foreign tenant, reject v3 over live broker",
  { skip: BOOTSTRAP ? false : "KAFKA_BOOTSTRAP not set" },
  async () => {
    // lazy import so the unit run (no kafkajs needed at module load) is unaffected
    const { Kafka, logLevel } = await import("kafkajs");

    const TENANT = "acme";
    const suffix = `it${Date.now().toString(36)}${Math.floor(Math.random() * 1e4)}`;
    const channel = `http-${suffix}`;
    const topic = `batch.task.dispatch.${TENANT}.${channel}`;
    const groupId = `${consumerGroupId(TENANT, "w-it")}-${suffix}`;

    const kafka = new Kafka({
      clientId: `it-producer-${suffix}`,
      brokers: [BOOTSTRAP!],
      logLevel: logLevel.NOTHING,
    });
    const admin = kafka.admin();
    const producer = kafka.producer();

    // collected pipeline outcomes, keyed by taskId for assertions
    const outcomes = new Map<string, PipelineOutcome>();
    const accepted: DispatchMessage[] = [];

    const adapter = new KafkaConsumerAdapter(
      {
        brokers: [BOOTSTRAP!],
        tenantId: TENANT,
        workerCode: "w-it",
        // PLAINTEXT locally; SASL/SCRAM-SHA-512 is config-supported (sasl:{...})
        groupId,
        fromBeginning: true,
      },
      silentLogger,
    );

    // sanity: the adapter targets exactly this tenant's wildcard + group
    assert.equal(adapter.groupId, groupId);
    assert.equal(adapter.topicRegex.source, dispatchTopicRegex(TENANT).source);
    assert.ok(adapter.topicRegex.test(topic), "regex should match the topic");

    const pipeline = new MessagePipeline({
      tenantId: TENANT,
      inFlight: () => 0,
      maxConcurrent: 8,
      assignment: assignmentOf(adapter),
      logger: silentLogger,
      onAccepted: async (msg) => {
        accepted.push(msg);
      },
    });

    // wrap pipeline.onMessage so we can record per-task outcomes for asserts,
    // while the adapter's runPipeline still owns the commit/pause policy.
    const recordingPipeline = {
      onMessage: async (r: { value: string }) => {
        const out = await pipeline.onMessage(r);
        try {
          const parsed = JSON.parse(r.value) as { taskId?: string };
          if (parsed.taskId) outcomes.set(parsed.taskId, out);
        } catch {
          /* parse-error path: no taskId to key on */
        }
        return out;
      },
    } as unknown as MessagePipeline;

    try {
      await admin.connect();
      await admin.createTopics({
        topics: [{ topic, numPartitions: 1, replicationFactor: 1 }],
        waitForLeaders: true,
      });
      await producer.connect();

      // start the adapter consuming with manual-commit policy (background)
      const running = adapter.runPipeline(recordingPipeline);

      // produce the three messages
      const validV1 = {
        schemaVersion: "v1",
        taskId: "task-valid-1",
        tenantId: TENANT,
        parameters: { foo: "bar" },
        runtimeAttributes: { traceId: "trace-it-1" },
      };
      const foreign = {
        schemaVersion: "v1",
        taskId: "task-foreign-1",
        tenantId: "other-tenant",
      };
      const v3 = {
        schemaVersion: "v3",
        taskId: "task-v3-1",
        tenantId: TENANT,
      };

      await producer.send({
        topic,
        messages: [
          { value: JSON.stringify(validV1) },
          { value: JSON.stringify(foreign) },
          { value: JSON.stringify(v3) },
        ],
      });

      // wait until all three have flowed through the pipeline
      await waitFor(() => outcomes.size >= 3);

      // 1. valid v1 → accepted + committed + handler saw it
      const acc = outcomes.get("task-valid-1");
      assert.ok(acc, "valid message should have an outcome");
      assert.equal(acc.kind, "accepted");
      assert.equal(acc.committed, true);
      assert.ok(
        accepted.some((m) => m.taskId === "task-valid-1"),
        "onAccepted should have received the valid v1 message",
      );

      // 2. foreign tenant → dropped, NOT committed (§1.9)
      const drop = outcomes.get("task-foreign-1");
      assert.ok(drop, "foreign message should have an outcome");
      assert.equal(drop.kind, "dropped-tenant");
      assert.equal(drop.committed, false);

      // 3. schemaVersion v3 → rejected, NOT committed (§A)
      const rej = outcomes.get("task-v3-1");
      assert.ok(rej, "v3 message should have an outcome");
      assert.equal(rej.kind, "rejected-schema");
      assert.equal(rej.committed, false);

      // foreign tenant must NEVER reach the handler
      assert.ok(
        !accepted.some((m) => m.tenantId !== TENANT),
        "no foreign-tenant message may reach the handler",
      );

      await Promise.race([
        running,
        adapter.wakeup().then(() => running),
      ]);
    } finally {
      await adapter.disconnect().catch(() => {});
      await producer.disconnect().catch(() => {});
      await admin.deleteTopics({ topics: [topic] }).catch(() => {});
      await admin.disconnect().catch(() => {});
    }
  },
);

File: sdk/go/kafka/kafka_consumer_integration_test.go
package kafka

import (
	"context"
	"encoding/json"
	"fmt"
	"log"
	"os"
	"strings"
	"testing"
	"time"

	"github.com/pinpols/file-batch-system/batch-worker-sdk-go/client"
	kgo "github.com/segmentio/kafka-go"
)

// Integration test for the real Kafka adapter. ENV-GATED: a plain `go test ./...`
// without KAFKA_BOOTSTRAP skips it, so the default test run is broker-free and
// stays green. Run against the local PLAINTEXT broker with:
//
//	KAFKA_BOOTSTRAP=localhost:19092 go test ./...
//
// SASL/SCRAM-SHA-512 + TLS are config-supported (see Config / buildDialer) but
// the local broker is PLAINTEXT, so this test exercises the PLAINTEXT path; the
// SCRAM path is validated by construction (buildDialer) and code review.
func TestKafkaConsumer_Integration(t *testing.T) {
	bootstrap := os.Getenv("KAFKA_BOOTSTRAP")
	if bootstrap == "" {
		t.Skip("KAFKA_BOOTSTRAP not set; skipping Kafka integration test (broker-free run)")
	}

	const tenant = "acme"
	// Unique suffix so concurrent / repeated runs don't collide.
	suffix := fmt.Sprintf("%d", time.Now().UnixNano())
	workerCode := "w-" + suffix
	// node-direct topic the consumer discovers by suffix (.node.<workerCode>):
	// batch.task.dispatch.<workerType>.node.<workerCode> (#2; was tenant-first).
	topic := fmt.Sprintf("batch.task.dispatch.http.node.%s", workerCode)

	brokers := strings.Split(bootstrap, ",")

	// --- create topic + cleanup -------------------------------------------
	createTopic(t, brokers[0], topic)
	t.Cleanup(func() { deleteTopic(brokers[0], topic) })

	// --- build the adapter under test -------------------------------------
	logger := log.New(os.Stderr, "[itest] ", log.LstdFlags)
	consumer, err := NewConsumer(Config{
		Brokers:    brokers,
		TenantID:   tenant,
		WorkerCode: workerCode,
		// Short refresh so the wildcard discovery is exercised quickly.
		TopicRefreshInterval: 2 * time.Second,
		FetchMaxWait:         500 * time.Millisecond,
		Logger:               logger,
	})
	if err != nil {
		t.Fatalf("NewConsumer: %v", err)
	}
	t.Cleanup(consumer.Close)

	// Pipeline bound to tenant "acme", maxConcurrent high so no backpressure.
	fsm := client.NewFSM()
	pipeline := client.NewMessagePipeline(tenant, fsm, 100, func() int { return 0 }, logger)

	// --- produce the three messages ---------------------------------------
	validMsg := client.TaskDispatchMessage{
		TaskID:        "task-" + suffix,
		TenantID:      tenant,
		SchemaVersion: "v1",
		WorkerType:    "http",
	}
	foreignMsg := client.TaskDispatchMessage{
		TaskID:        "foreign-" + suffix,
		TenantID:      "other-tenant",
		SchemaVersion: "v1",
		WorkerType:    "http",
	}
	badSchemaMsg := client.TaskDispatchMessage{
		TaskID:        "badschema-" + suffix,
		TenantID:      tenant,
		SchemaVersion: "v3",
		WorkerType:    "http",
	}
	produce(t, brokers, topic, validMsg, foreignMsg, badSchemaMsg)

	// --- consume + run through the pipeline -------------------------------
	// Collect dispositions for the three known taskIds. We poll until we have
	// observed all three or hit a deadline.
	deadline := time.Now().Add(30 * time.Second)
	seen := map[string]client.MessageDisposition{}
	wantIDs := map[string]bool{
		validMsg.TaskID:     true,
		foreignMsg.TaskID:   true,
		badSchemaMsg.TaskID: true,
	}

	for time.Now().Before(deadline) && len(seen) < len(wantIDs) {
		recs, ok := consumer.Poll()
		if !ok {
			t.Fatalf("consumer Poll returned not-ok unexpectedly")
		}
		for _, rec := range recs {
			msg, disp := pipeline.OnMessage(rec)
			if !wantIDs[msg.TaskID] {
				continue // ignore stray records from other runs
			}
			seen[msg.TaskID] = disp
			// Commit ONLY on acceptance; otherwise drop pending so the offset
			// is withheld and the message would be redelivered.
			if disp == client.DispositionAccepted {
				consumer.Commit()
			} else {
				consumer.DropPending()
			}
		}
	}

	// --- assertions --------------------------------------------------------
	if got := seen[validMsg.TaskID]; got != client.DispositionAccepted {
		t.Errorf("valid v1 message: disposition = %q, want ACCEPTED", got)
	}
	if got := seen[foreignMsg.TaskID]; got != client.DispositionDroppedForeignTenant {
		t.Errorf("foreign-tenant message: disposition = %q, want DROPPED_FOREIGN_TENANT", got)
	}
	if got := seen[badSchemaMsg.TaskID]; got != client.DispositionRejectedSchema {
		t.Errorf("schemaVersion=v3 message: disposition = %q, want REJECTED_SCHEMA", got)
	}
	if len(seen) != len(wantIDs) {
		t.Errorf("only saw %d/%d expected messages: %v", len(seen), len(wantIDs), seen)
	}
}

// createTopic creates a single-partition topic via the controller broker.
func createTopic(t *testing.T, broker, topic string) {
	t.Helper()
	conn, err := kgo.Dial("tcp", broker)
	if err != nil {
		t.Fatalf("dial broker: %v", err)
	}
	defer conn.Close()

	controller, err := conn.Controller()
	if err != nil {
		t.Fatalf("get controller: %v", err)
	}
	ctrlConn, err := kgo.Dial("tcp", fmt.Sprintf("%s:%d", controller.Host, controller.Port))
	if err != nil {
		t.Fatalf("dial controller: %v", err)
	}
	defer ctrlConn.Close()

	if err := ctrlConn.CreateTopics(kgo.TopicConfig{
		Topic:             topic,
		NumPartitions:     1,
		ReplicationFactor: 1,
	}); err != nil {
		t.Fatalf("create topic %s: %v", topic, err)
	}

	// Wait until metadata reports the topic so the GroupReader can find it.
	wait := time.Now().Add(15 * time.Second)
	for time.Now().Before(wait) {
		parts, err := conn.ReadPartitions(topic)
		if err == nil && len(parts) > 0 {
			return
		}
		time.Sleep(300 * time.Millisecond)
	}
	t.Fatalf("topic %s did not appear in metadata", topic)
}

// deleteTopic removes the test topic (best-effort cleanup).
func deleteTopic(broker, topic string) {
	conn, err := kgo.Dial("tcp", broker)
	if err != nil {
		return
	}
	defer conn.Close()
	controller, err := conn.Controller()
	if err != nil {
		return
	}
	ctrlConn, err := kgo.Dial("tcp", fmt.Sprintf("%s:%d", controller.Host, controller.Port))
	if err != nil {
		return
	}
	defer ctrlConn.Close()
	_ = ctrlConn.DeleteTopics(topic)
}

// produce writes the given dispatch messages as JSON to the topic.
func produce(t *testing.T, brokers []string, topic string, msgs ...client.TaskDispatchMessage) {
	t.Helper()
	w := &kgo.Writer{
		Addr:         kgo.TCP(brokers...),
		Topic:        topic,
		Balancer:     &kgo.LeastBytes{},
		BatchTimeout: 100 * time.Millisecond,
	}
	defer w.Close()

	var records []kgo.Message
	for _, m := range msgs {
		b, err := json.Marshal(m)
		if err != nil {
			t.Fatalf("marshal msg: %v", err)
		}
		records = append(records, kgo.Message{Key: []byte(m.TaskID), Value: b})
	}
	ctx, cancel := context.WithTimeout(context.Background(), 15*time.Second)
	defer cancel()
	if err := w.WriteMessages(ctx, records...); err != nil {
		t.Fatalf("produce: %v", err)
	}
}
